transcom/mymove
# Prime API BC - B 18107 shipment diversion

## [B-18107](https://www13.v1host.com/USTRANSCOM38/story.mvc/Summary?oidToken=Story%3A871300&RoomContext=TeamRoom%3A808731&concept=TeamRoom)

## Summary

### Breaking API Change
`POST createMTOShipment` is to be deprecated under prime V1 and be replaced by V2, with a sunset date of April 8th, 2024.

This POST endpoint now requires the parameter `divertedFromShipmentId` to be present if the `diversion` parameter is present and set to `true`

### Service logic

I refactored the mto_shipment service for createMtoShipment to allow validators to be passed in on creation, of which v1 now has the `protectV1Diversion` rule and v2 has the `checkDiversionValid` rule.
## Verification Steps for Reviewers

These are to be checked by a reviewer.

- [x] Are you correctly using the prime/v1/ or prime/v2/ endpoint during postman testing?

### New Test Coverage

When reimplementing existing V1 functionality for the Prime under V2, I found that the service objects and models used had no tests. Included in this PR is a backfill of tests for V1 and V2 to satisfy code coverage requirements.

### Setup to Run the Code

- [Instructions for starting storybook](https://transcom.github.io/mymove-docs/docs/frontend/setup/storybook)
- [Instructions for starting the MilMove application](https://transcom.github.io/mymove-docs/docs/getting-started/application-setup/)
- [Instructions for running tests](https://transcom.github.io/mymove-docs/docs/getting-started/development/testing)

### How to test

1. Boot up postman
2. Import the new Prime API V2 .yaml file
3. Navigate to the new POST endpoint
  a. <img width="251" alt="image" src="https://github.com/transcom/mymove/assets/136814362/f49d4543-6d9c-496e-b109-e18943e43d73"> 
4. Make sure to update your environment variable to have `prime/v2/`, `{{baseUrl}}` is set by default to `prime/v1/`
5. Use the example payload
6. Modify payload accordingly for testing
7. Review in database under `mto_shipments` accordingly
 
Example payload

```
{
  "moveTaskOrderID": "59476e8f-5efe-407a-8da8-22d04d185a7d",
  "requestedPickupDate": "2023-12-19",
  "primeEstimatedWeight": 4500,
  "customerRemarks": "handle with care",
  "agents": [
    {
      "firstName": "string",
      "lastName": "string",
      "email": "I2Zo35dd%6hSYW2Xfi6wFD+G6CDGh44CfcWK2.ypLOWMlMTTNOKw@RmRi2KKDbC4to.yvLbJPMGTJChrkTbdshELMiTYErLIELVrjknRYzMlbBVNnUGdRYmMXELYp",
      "phone": "",
      "agentType": "RELEASING_AGENT"
    }
  ],
  "pickupAddress": {
    "streetAddress1": "123 Main Ave",
    "streetAddress2": "Apartment 9000",
    "streetAddress3": "Montmârtre",
    "city": "Anytown",
    "state": "AL",
    "postalCode": "90210",
    "country": "USA"
  },
  "shipmentType": "HHG",
  "diversion": true,
  "divertedFromShipmentId": "8b646662-a95a-41c6-a0f0-b3ab157a1935",
  "pointOfContact": "string",
  "counselorRemarks": "counselor approved",
  "ppmShipment": {
    "expectedDepartureDate": "2023-12-19",
    "pickupPostalCode": "90210",
    "secondaryPickupPostalCode": "90210",
    "destinationPostalCode": "90210",
    "secondaryDestinationPostalCode": "90210",
    "sitExpected": true,
    "sitLocation": "ORIGIN",
    "sitEstimatedWeight": 2000,
    "sitEstimatedEntryDate": "2023-12-19",
    "sitEstimatedDepartureDate": "2023-12-19",
    "estimatedWeight": 4200,
    "hasProGear": true,
    "proGearWeight": 0,
    "spouseProGearWeight": 0
  }
}
```
